reverseforwardrecord/get_dns_record.py

```diff
@@ -0,0 +1,255 @@
+    :param fqdn_column: full qualified domain name column in netbox file
+    :param idrac_ip_column: idrac ip column name in netbox file
+    :return: a list of dictionaries containing fqdn and idrac_ip pairs
+    """
+
+    sheetnames = pd.ExcelFile(netbox_filepath).sheet_names
+
+    # get all sheets on the XLSX file.
+    workbook_info = []
+
+    for i, sheet in enumerate(sheetnames, 1):
+        workbook = pd.read_excel(netbox_filepath, sheet_name=sheet, engine='openpyxl').to_dict()
+
+        if fqdn_column not in workbook or idrac_ip_column not in workbook:
+            raise RuntimeError(
+                f"Error: you've provided an excel sheet: (no. {i}) which does not contain one or more of these "
+                f"following column in workbook: {fqdn_column}, {idrac_ip_column}"
+            )
+        for key, fqdn_val in workbook[fqdn_column].items():
+            workbook_info.extend([
+                {
+                    "fqdn": fqdn_val,
+                    "idrac_ip": workbook[idrac_ip_column][key]
+                }
+            ])
+    return workbook_info
+
+
+def parse_netbox_info(
+        netbox_info: List[Dict], reverse_order: bool = False
+) -> List[Dict]:
+    """
+    Parse info from netbox and create corresponding hypervisor fqdn and ip value
+    :param netbox_info: list of dictionaries holding fqdn and idrac_ip
+    :param reverse_order: use reverse or forward order / if true use forward if  false use reverse
+    :return: list of dictionaries which will contain hypervisor fqdn and ip keys-value pairs
+    """
+    parsed_info = []
+    for item in netbox_info:
+        parsed_item = {}
+        if reverse_order:
+            idrac = reduce(
+                lambda x, y: f"{x}.{y}", item["idrac_ip"].split(".")[::-1]
+            )  #
+            idrac = idrac.rsplit(".", 1)[0]
+            parsed_item["ip_address"] = idrac
+            parsed_item["hypervisor"] = item["fqdn"]
+        else:
+            parsed_item["ip_address"] = item["idrac_ip"]
+            parsed_item["hypervisor"] = item["fqdn"].split(".")[0]
+        parsed_info.append(parsed_item)
+    return parsed_info
+
+
+def write_output(
+        parsed_info: List[Dict], output_filepath: str, reverse_order: bool = False
+):
+    """
+    write parsed netbox info to a file
+    :param parsed_info: list of dictionaries which contains hypervisor fqdn and ip keys
+    :param output_filepath: file path for output
+    :param reverse_order: use reverse or forward order / if true use forward if false use reverse
+    :return: the text written into the file
+    """
+    print(output_filepath)
+    with open(output_filepath, "w", encoding="utf-8") as out_file:
+        for item in parsed_info:
+            if reverse_order:
+                out_file.write(f"{item['ip_address']}\tIN PTR\t{item['hypervisor']}\n")
+            else:
+                out_file.write(f"{item['hypervisor']}\tIN A\t{item['ip_address']}\n")
```

Comment:
whenever you open() a file to write to it we prohibit other processes from accessing that file. Therefore we need to be considerate to others and ensure that we process the file as quick as possible. 
Here, we are doing some processing WHILST the file is open - which can be done before we open the file - making writing quicker. Here's an example of how we can make this quicker: 

```suggestion
   lines = []
   for item in parsed_info:
            if reverse_order:
                lines.append(f"{item['ip_address']}\tIN PTR\t{item['hypervisor']}\n")
            else:
                lines.append(f"{item['hypervisor']}\tIN A\t{item['ip_address']}\n")

    with open(output_filepath, "w", encoding="utf-8") as out_file:
        outfile.writelines(lines)
```

Reply:
very helpful thank you